qiskit_dynamics/solvers/solver_classes.py

```diff
@@ -503,187 +515,118 @@ def solve(
-                self.model.__class__,
-                self._hamiltonian_channels,
-                self._dissipator_channels,
-            )
-
-            self._set_new_signals(signals)
+        return all_results if multiple_sims else all_results[0]
 
-            results = solve_lmde(generator=self.model, t_span=t_span, y0=y0, **kwargs)
-            results.y = format_final_states(results.y, self.model, y0_input, y0_cls)
-
-            # reset signals to ensure purity
-            self.model.signals = model_sigs
+    @contextmanager
+    def signal_assignment(self, signals: Union[Tuple[List[Signal], List[Signal]], List[Signal]]):
+        """Context manager with temporarily changes the model's signals to those provided
+        and changes them back when the context is exited."""
+        old_signals = self.model.signals
 
-            return Array(results.t).data, Array(results.y).data
+        # if Lindblad model and signals are given as a list set as Hamiltonian part of signals
+        if isinstance(self.model, LindbladModel) and isinstance(signals, (list, SignalList)):
+            signals = (signals, None)
 
-        jit_sim_function = jit(sim_function, static_argnums=(4,))
-
-        # run simulations
-        all_results = []
-        for t_span, y0, sched in zip(t_span_list, y0_list, schedule_list):
-            # setup initial state
-            y0, y0_input, y0_cls, state_type_wrapper = validate_and_format_initial_state(
-                y0, self.model
-            )
+        if self._rwa_signal_map:
+            signals = self._rwa_signal_map(signals)
+        self.model.signals = signals
 
-            # setup array of all samples
-            all_signals = self._schedule_converter.get_signals(sched)
+        yield
 
-            all_samples = np.zeros(all_samples_shape, dtype=complex)
-            for idx, sig in enumerate(all_signals):
-                all_samples[idx, 0 : len(sig.samples)] = np.array(sig.samples)
+        self.model.signals = old_signals
 
-            results_t, results_y = jit_sim_function(
-                Array(t_span).data, Array(y0).data, all_samples, Array(y0_input).data, y0_cls
-            )
-            results = OdeResult(t=results_t, y=Array(results_y, backend="jax", dtype=complex))
-
-            if y0_cls is not None and convert_results:
-                results.y = [state_type_wrapper(yi) for yi in results.y]
-
-            all_results.append(results)
+    def _jit_shape(self, required_duration: int):
+        """If possible, return a nearby samples shape so that we don't have to recompile. Otherwise,
+        add a new one to the cache."""
+        for duration in sorted(self._jit_cache):
+            if duration * self._JIT_MIN_PADDING_FRAC < required_duration <= duration:
+                # something appropriate was found in the cache
+                break
+        else:
+            duration = round(self._JIT_DURATION_PADDING * required_duration)
+            self._jit_cache.add(duration)
```

Comment:
Looks like these lines should be at one level of indentation deeper.

Reply:
this is a for-else statement so i think its correct. one python designer once quipped that he wishes it had been named for-nobreak, because everyone always gets confused about the meaning of else there